### ISSUE TITLE
Add Dockerfile workspace information as Metadata

### ISSUE THREAD

**Issue by alexschroeter** [alexschroeter/apptainer-deployer#3]
I came across the problem that the Dockerfile `workspace` setting is not propagated to Apptainer. A maybe not very nice but simple solution would be to put this information somewhere else for Apptainer to use it.

**Comment by jhnnsrs** [comment]
hmm, that sounds a bit annoying and would negate a bit the principle of not interfering with the containering software itself (ie. this is the docker/apptainer part). However we could have maybe the command run in the  container itself i.e ( CMD "arkitekt-next run prod" and just pass the FAKTS token as an environment variable. This would allow the container to decide which and where to run the arkitekt-next command itself (not sure if this works with apptainer though).

TLDR: instead of calling arkitekt-next from the deployer app use "arkitekt-next" as a cmd inside the container and read out the necessary variables from the environment?"

**Comment by alexschroeter** [comment]
My initial thought was doing something like this where `/app` directory comes from the metadata that is written in the manifest.yml, but I see your point.
```
aschroeter@Watson:~$ export APPTAINER_TARGET_PWD=/app
aschroeter@Watson:~$ apptainer shell docker://jhnnsrs/renderer:0.0.1-vanilla
INFO:    Using cached SIF image
INFO:    squashfuse not found, will not be able to mount SIF or other squashfs files
INFO:    fuse2fs not found, will not be able to mount EXT3 filesystems
INFO:    gocryptfs not found, will not be able to use gocryptfs
INFO:    Converting SIF file to temporary sandbox...
Apptainer> pwd
/app
```

If I understand you correctly, you would have `arkitekt-next run prod` read the `/app` directory (from where? The Dockerfile? Or also Metadata? Or are you saying Fakts does already know this?) and switch to that folder? And because it is bundled inside the container, there is no dependency to how it is started and able to run independently. Makes sense, I think.

**Comment by alexschroeter** [comment]
Created a ticket https://github.com/apptainer/apptainer/issues/2573

**Comment by alexschroeter** [comment]
created a ticket to check if this behavior is correct https://github.com/apptainer/apptainer/issues/2573

**Comment by alexschroeter** [comment]
It is confirmed in the ticket that it is intended behavior, and there is no way to get that information.

To recap, I see these 2 options:

1. We could add docker inspect to get the working directory when the `ENTRYPOINT ["arkitekt-next"]` will start to check for the right directory (this would mean all apps need docker/skopeo to perform this)

I tried a prototype for this but found that docker inspect doesn't run without a docker daemon running. I found [skopeo](https://github.com/containers/skopeo/blob/main/README.md) does this and created a working example based on the example I used for the [bug report](https://github.com/apptainer/apptainer/issues/2573) with pwd where we would run `arkitekt-next` instead.

Here is the result:
```bash 
docker run alexanderschroeter/workdir-test:workaround
/myworkdir

apptainer run docker://alexanderschroeter/workdir-test:workaround
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Copying blob bad8b0455b54 done   |
Copying blob f10730a567f7 done   |
Copying blob ff65ddf9395b skipped: already exists
Copying blob 773c2b8def60 done   |
Copying config 4cf9044dd0 done   |
Writing manifest to image destination
2024/10/18 22:06:58  info unpack layer: sha256:ff65ddf9395be21bfe1f320b7705e539ee44c1053034f801b1a3cbbf2d0f4056
2024/10/18 22:06:59  info unpack layer: sha256:f10730a567f7d73f6b2dc925004aaf2b17b722b3322004fce1b77b72c5dc1600
2024/10/18 22:07:01  info unpack layer: sha256:773c2b8def60f2a36d8f530aa49edb08f189a0498540e93458783b0ae3af7b4f
2024/10/18 22:07:01  info unpack layer: sha256:bad8b0455b542a2480e22bf0a76a75c1ee458c42273b758f7ed3faad7bb751f2
INFO:    Creating SIF file...
INFO:    squashfuse not found, will not be able to mount SIF or other squashfs files
INFO:    fuse2fs not found, will not be able to mount EXT3 filesystems
INFO:    gocryptfs not found, will not be able to use gocryptfs
INFO:    Converting SIF file to temporary sandbox...
WARN[0000] "/mnt/wslg/runtime-dir" directory set by $XDG_RUNTIME_DIR does not exist. Either create the directory or unset $XDG_RUNTIME_DIR.: stat /mnt/wslg/runtime-dir: no such file or directory: Trying to pull image in the event that it is a public image.
/myworkdir
INFO:    Cleaning up image...
```

2. or put the working directory somewhere inside the container during the build so the `ENTRYPOINT ["arkitekt-next"]` will change directory to the working directory before executing



I am not a huge fan of the overhead which we would need to add to all images and would probably prefer adding the working directory to some metadata during the `arkitekt build` phase. All in all, I am not very happy with this but the only other way to solve this would probably be generating the Apptainer equivalent of the Dockerfile for all Apps which seems even worse. I am a bit frustrated with these options. :(

EDIT:
Wait, we don't need to add inspect to all containers. It should be enough for apptainer-deployer to do the inspection and pass it to the image. I think I made it more complicated than it needed to be. This would still mean that the docker containers work independent, but Apptainer only works with the apptainer-deployer. Which I can live with, since I don't see a future where Apptainer would be the preferred way of doing things, but we have support for it.

I will try to find a python depency for the apptainer-deployer that does what skopeo does, since I don't want a dependency on installing something. Maybe there is a python module or I will just create a container that we can use to perform this inspect. Add it to the parameters for Apptainer and be done with it. What do you think?

**Comment by alexschroeter** [comment]
Solved by inspecting the dockerhub information for the workdir information which is passed to apptainer-deployer as parameter.

https://github.com/alexschroeter/apptainer-deployer/blob/2ea7dbbc3348d217bc4280ec693c28d81803dd6d/app.py#L271C1-L276C3
